### PR TITLE
Update bokehjs' plotting API to match bokeh

### LINKED ISSUE #11016: [BUG] Figure name passed to `renderer.glyph.name` but not `renderer.name`
Version: 2.2.3 and 2.3
Browsers: Chrome, Brave, Safari

When creating a new plot with the `name` property set, the given name is assigned to `renderer.glyph.name` but not passed to `renderer.name`, which is read by `HoverTool()`. This means this name will not be visible via the special field `$name` in `HoverTool`, as expected.

Key source code (key lines)

```
var plot = new Bokeh.Plotting.figure()
plot.line([0,1],[0,1],{name:"the_name"})
```

The above code will produce the following output:

```
> plot.renderers[0].name
null
>plot.renderers[0].glyph.name
"the_name"
```

A complete, minimal example showing the original bug and a workaround is given via CodePen at [bokehjs_glyph_name](https://codepen.io/oviquezr/pen/PobOKoo?editors=1111). The workaround simply consists of passing the glyph name to the renderer name, or otherwise setting the renderer name manually.

Issue is related to discourse topic: [Differences in BokehJS vs Python : expected usage](https://discourse.bokeh.org/t/differences-in-bokehjs-vs-python-expected-usage/7313)

### LINKED ISSUE #11035: [BUG] `Bokeh.Widgets` objects missing `tools`, required by `Bokeh.Plotting.gridplot()`
Version: 2.2.3 and 2.3
Browsers: Chrome, Brave, Safari

A `Div()` instance created by calling `var mydiv = new Bokeh.Widgets.Div()` does not include `mydiv.toolbar.tools` to facilitate using in a custom `Bokeh.Plotting.gridplot()`; or `gridplot()` fails to check whether the property exists before accessing. This throws:

```
Uncaught TypeError: Cannot read property 'tools' of undefined
```

To use a `Div()` inside a `gridplot()`, my current workaround is to first set the div's toolbar property, as follows:

```
mydiv.toolbar = {tools:[]}
```

A complete, minimal example is given via CodePen at [bokehjs_div_tools](https://codepen.io/oviquezr/pen/gOLzbvZ), where JS:line 17 contains the workaround noted above. Comment this line out to trigger the failure.

Issue is related to discourse topic: [Differences in BokehJS vs Python : expected usage](https://discourse.bokeh.org/t/differences-in-bokehjs-vs-python-expected-usage/7313)

### LINKED ISSUE #11037: [BUG] Using `sizing_mode` in gridplot layouts requires explicit assignment
Version: 2.2.3 and 2.3
Browsers: Chrome, Brave, Safari

`Bokeh.Plotting.gridplot()` does not inherit `sizing_mode` from the underlying plots. To match the expected behavior from the Python library, for example when using `sizing_mode: 'stretch_width'`, the `sizing_mode` must be explicitly stated for both the `figure()` and the `gridplot()` objects.

A complete, minimal example is given via CodePen at [bokehjs_sizing_mode](https://codepen.io/oviquezr/pen/eYBRjXL). The example shows two instances of `gridplot`, each with one `sizing_mode:'fixed'` and one `sizing_mode:'stretch_width'`. Only the second `gridplot` is given an explicit `sizing_mode:'stretch_width'` and adjusts to its container correctly.

Issue is related to discourse topic: [Differences in BokehJS vs Python : expected usage](https://discourse.bokeh.org/t/differences-in-bokehjs-vs-python-expected-usage/7313)

### LINKED ISSUE #11038: [BUG] Calling `figure({title:"some title"})` replaces Title object with string, prevents subsequent updates to title text
Version: 2.2.3 and 2.3
Browsers: Chrome, Brave, Safari

When creating a figure with `var plot = new Bokeh.Plotting.figure()`, the property `plot.title` will instantiate from `Bokeh.Title()`. However, when calling `var plot = new Bokeh.Plotting.figure({title:"some title"})`, BokehJS replaces `plot.title` with the provided string. This makes it so that the title text cannot be dynamically updated later. In Python, the
equivalent approach casts the string in the argument to the underlying Title() object.

To preserve the ability to update a figure title later, the following sequence works as expected:

```
var plot = new Bokeh.Plotting.figure()
plot.title.text = "some title"
```

A complete, minimal example is given via CodePen at [bokehjs_titles](https://codepen.io/oviquezr/pen/LYbLBJK). The example shows two figures, each using one of the methods described above. The example also provides a button to trigger an update to both figure titles; however, only one of the titles can be updated as needed. JS lines 44-51 show the update function:

```
function addTitle() {
  var new_val = Math.floor(Math.random()*10)
  // plot1 title is string
  plot1.title = 'New Title '+new_val
  
  //plot2 title is Title object
  plot2.title.text = 'New Title '+new_val+' #2'
}
```

Note that `plot1.title` in this example is of type string, as set by JS lines:7-12 so it does **not** contain a `.text` value, while `plot2.title` is an instance of `Bokeh.Title` and thus provides the property `plot2.title.text`.

Issue is related to discourse topic: [Differences in BokehJS vs Python : expected usage](https://discourse.bokeh.org/t/differences-in-bokehjs-vs-python-expected-usage/7313)

### PR DESCRIPTION
**Pull request by mattpap**
fixes #11016
fixes #11035
fixes #11037
fixes #11038

### CODE REVIEW

File: bokehjs/src/lib/models/plots/plot.ts [hunk 1]
```diff
@@ -135,7 +135,10 @@ export class Plot extends LayoutDOM {
       frame_width:       [ Nullable(Number), null ],
       frame_height:      [ Nullable(Number), null ],
 
-      title:             [ Or(Ref(Title), String, Null), () => new Title({text: ""}) ],
+      // revise this when https://github.com/microsoft/TypeScript/pull/42425 is merged
+      title:             [ Or(Ref(Title), String, Null), "", {
+        convert: (title) => isString(title) ? new Title({text: title}) : title,
```
Comment (bryevdv) [inline bokehjs/src/lib/models/plots/plot.ts:140]: why not `accepts` to mirror the Python side?

File: bokehjs/test/unit/core/util/pretty.ts [hunk 2]
```diff
@@ -38,5 +38,12 @@ describe("core/util/pretty module", () => {
       expect(to_string({})).to.be.equal("{}")
       expect(to_string(new Map([[1, true], [2, false], [3, true]]))).to.be.equal("Map([[1, true], [2, false], [3, true]])")
     })
+
+    it("that supports curcular objects", () => {
```
Comment (bryevdv) [inline bokehjs/test/unit/core/util/pretty.ts:42]:
```suggestion
    it("that supports circular objects", () => {
```

File: bokehjs/src/lib/core/util/pretty.ts [hunk 3]
```diff
@@ -18,12 +18,20 @@ export type PrinterOptions = {
...
+  protected readonly visited: Set<object> = new Set()
 
   constructor(options?: PrinterOptions) {
     this.precision = options?.precision
   }
 
   to_string(obj: unknown): string {
+    if (isObject(obj)) {
+      if (this.visited.has(obj))
+        return "<circular>"
```
Comment (bryevdv) [inline bokehjs/src/lib/core/util/pretty.ts:30]: slight preference to `<cycle>`  here (noun instead of adjective)

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Some tiny comments other wise LGTM

File: bokehjs/src/lib/core/util/pretty.ts (hunk 3 above)
Comment (mattpap) [reply bokehjs/src/lib/core/util/pretty.ts:30]: I'm using node's nomenclature here:
```
$ node
Welcome to Node.js v14.14.0.
Type ".help" for more information.
> x = {y: 1}
{ y: 1 }
> x.z = x
<ref *1> { y: 1, z: [Circular *1] }
```

File: bokehjs/src/lib/models/plots/plot.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/models/plots/plot.ts:140]: I didn't think about Python here, but also it has different semantics.

**Pull request merged**